File: src/spinehub/utils/slack-channels.ts
/**
 * Slack Channel Utils - Channel to theme mapping
 *
 * Provides utilities for:
 * - Mapping Slack channels to worklog themes
 * - Channel categorization
 * - Theme inference from channel activity
 */

import { ChannelMapping } from '../types';

// Re-export types
export { ChannelMapping };

/**
 * Known channel mappings for TSA projects
 */
export const CHANNEL_THEME_MAP: Record<string, { theme: string; category: string }> = {
  // Customer Projects
  'intuit-internal': { theme: 'Intuit WFS Project', category: 'customer' },
  'testbox-intuit-wfs-external': { theme: 'WFS External Coordination', category: 'customer' },
  'testbox-intuit-mailchimp-external': { theme: 'Mailchimp Integration', category: 'customer' },
  'external-testbox-apollo': { theme: 'Apollo Project', category: 'customer' },
  'brevo-internal': { theme: 'Brevo Integration', category: 'customer' },
  'archer-internal': { theme: 'Archer Project', category: 'customer' },

  // Team Channels
  'tsa-data-engineers': { theme: 'TSA Team Sync', category: 'team' },
  'team-koala': { theme: 'Koala Team', category: 'team' },
  'dev-on-call': { theme: 'Engineering Support', category: 'ops' },

  // Company Channels
  'product': { theme: 'Product Discussions', category: 'internal' },
  'go-to-market': { theme: 'GTM Strategy', category: 'internal' },
  'engineering': { theme: 'Engineering', category: 'internal' },
  'general': { theme: 'General', category: 'internal' },
};

/**
 * Get theme for a channel
 */
export function getChannelTheme(channelName: string): string {
  // Direct match
  if (CHANNEL_THEME_MAP[channelName]) {
    return CHANNEL_THEME_MAP[channelName].theme;
  }

  // Partial match
  const lowerChannel = channelName.toLowerCase();
  for (const [pattern, mapping] of Object.entries(CHANNEL_THEME_MAP)) {
    if (lowerChannel.includes(pattern.toLowerCase())) {
      return mapping.theme;
    }
  }

  // Default to channel name
  return `Channel: #${channelName}`;
}

/**
 * Get category for a channel
 */
export function getChannelCategory(channelName: string): string {
  // Direct match
  if (CHANNEL_THEME_MAP[channelName]) {
    return CHANNEL_THEME_MAP[channelName].category;
  }

  // Partial match
  const lowerChannel = channelName.toLowerCase();
  for (const [pattern, mapping] of Object.entries(CHANNEL_THEME_MAP)) {
    if (lowerChannel.includes(pattern.toLowerCase())) {
      return mapping.category;
    }
  }

  // Infer category from name patterns
  if (lowerChannel.includes('external') || lowerChannel.includes('client')) {
    return 'customer';
  }
  if (lowerChannel.includes('team') || lowerChannel.includes('squad')) {
    return 'team';
  }
  if (lowerChannel.includes('oncall') || lowerChannel.includes('ops') || lowerChannel.includes('incident')) {
    return 'ops';
  }

  return 'internal';
}

/**
 * Get full channel mapping
 */
export function getChannelMapping(channelId: string, channelName: string): ChannelMapping {
  return {
    channel_id: channelId,
    channel_name: channelName,
    theme: getChannelTheme(channelName),
    category: getChannelCategory(channelName),
  };
}

/**
 * Group channels by category
 */
export function groupChannelsByCategory(
  channels: Array<{ id: string; name: string }>
): Record<string, ChannelMapping[]> {
  const grouped: Record<string, ChannelMapping[]> = {
    customer: [],
    team: [],
    ops: [],
    internal: [],
  };

  for (const channel of channels) {
    const mapping = getChannelMapping(channel.id, channel.name);
    if (!grouped[mapping.category]) {
      grouped[mapping.category] = [];
    }
    grouped[mapping.category].push(mapping);
  }

  return grouped;
}

/**
 * Infer theme from channel content
 */
export function inferThemeFromContent(channelName: string, sampleMessages: string[]): string {
  const combinedText = sampleMessages.join(' ').toLowerCase();

  // Check for specific keywords
  if (combinedText.includes('release') || combinedText.includes('deploy')) {
    return `${getChannelTheme(channelName)} - Release`;
  }
  if (combinedText.includes('bug') || combinedText.includes('fix') || combinedText.includes('issue')) {
    return `${getChannelTheme(channelName)} - Bug Investigation`;
  }
  if (combinedText.includes('meeting') || combinedText.includes('sync') || combinedText.includes('call')) {
    return `${getChannelTheme(channelName)} - Coordination`;
  }
  if (combinedText.includes('document') || combinedText.includes('spec') || combinedText.includes('doc')) {
    return `${getChannelTheme(channelName)} - Documentation`;
  }
  if (combinedText.includes('review') || combinedText.includes('pr') || combinedText.includes('merge')) {
    return `${getChannelTheme(channelName)} - Code Review`;
  }

  return getChannelTheme(channelName);
}

/**
 * Check if channel is customer-facing
 */
export function isCustomerChannel(channelName: string): boolean {
  return getChannelCategory(channelName) === 'customer';
}

/**
 * Check if channel is internal
 */
export function isInternalChannel(channelName: string): boolean {
  return getChannelCategory(channelName) === 'internal';
}

/**
 * Get priority of channel for worklog ordering
 */
export function getChannelPriority(channelName: string): number {
  const category = getChannelCategory(channelName);

  // Priority: customer > ops > team > internal
  const priorities: Record<string, number> = {
    customer: 1,
    ops: 2,
    team: 3,
    internal: 4,
  };

  return priorities[category] || 5;
}

/**
 * Sort channels by worklog priority
 */
export function sortChannelsByPriority(channels: string[]): string[] {
  return [...channels].sort((a, b) => {
    return getChannelPriority(a) - getChannelPriority(b);
  });
}

File: python/utils/slack_channels.py
"""
Slack Channel Utils - Channel to theme mapping.

Provides utilities for:
- Mapping Slack channels to worklog themes
- Channel categorization
- User role inference

Ported from TypeScript: src/spinehub/utils/slack-channels.ts
Used by bridge.py handle_slack() handler.
"""

from dataclasses import dataclass, field
from typing import Optional


@dataclass
class ChannelInfo:
    """Basic Slack channel information."""
    id: str
    name: str
    is_private: bool = False
    num_members: int = 0


@dataclass
class ChannelMappingResult:
    """Result of mapping a channel to TSA context."""
    channel: str
    channel_id: str
    tsa_members: list[str] = field(default_factory=list)
    eng_members: list[str] = field(default_factory=list)
    gtm_members: list[str] = field(default_factory=list)
    external_members: list[str] = field(default_factory=list)


# Known channel mappings for TSA projects
CHANNEL_THEME_MAP: dict[str, dict[str, str]] = {
    # Customer Projects
    "intuit-internal": {"theme": "Intuit WFS Project", "category": "customer"},
    "testbox-intuit-wfs-external": {"theme": "WFS External Coordination", "category": "customer"},
    "testbox-intuit-mailchimp-external": {"theme": "Mailchimp Integration", "category": "customer"},
    "external-testbox-apollo": {"theme": "Apollo Project", "category": "customer"},
    "brevo-internal": {"theme": "Brevo Integration", "category": "customer"},
    "archer-internal": {"theme": "Archer Project", "category": "customer"},
    # Team Channels
    "tsa-data-engineers": {"theme": "TSA Team Sync", "category": "team"},
    "team-koala": {"theme": "Koala Team", "category": "team"},
    "dev-on-call": {"theme": "Engineering Support", "category": "ops"},
    # Company Channels
    "product": {"theme": "Product Discussions", "category": "internal"},
    "go-to-market": {"theme": "GTM Strategy", "category": "internal"},
    "engineering": {"theme": "Engineering", "category": "internal"},
    "general": {"theme": "General", "category": "internal"},
}

# Known TSA team roles
TSA_ROLES: dict[str, str] = {
    "thiago": "tsa_lead",
    "diego": "tsa",
    "gabrielle": "tsa",
    "carlos": "tsa",
    "alexandra": "tsa",
    "lucas": "engineering",
    "sam": "engineering",
    "waki": "engineering",
}


class SlackChannelMapper:
    """Maps Slack channels to worklog themes and categories."""

    def __init__(
        self,
        token: Optional[str] = None,
        company_domain: str = "@testbox.com",
        channel_prefixes: Optional[list[str]] = None,
    ):
        self.token = token
        self.company_domain = company_domain
        self.channel_prefixes = channel_prefixes or [
            "intuit", "testbox", "brevo", "archer",
            "tsa", "team", "dev", "product", "go-to-market",
        ]

    def list_channels(self, prefixes: Optional[list[str]] = None) -> list[ChannelInfo]:
        """
        List channels matching prefixes.
        Note: Without a live Slack token, returns known channels from theme map.
        """
        target_prefixes = prefixes or self.channel_prefixes
        channels: list[ChannelInfo] = []

        for channel_name, mapping in CHANNEL_THEME_MAP.items():
            if any(channel_name.startswith(p) or p in channel_name for p in target_prefixes):
                channels.append(ChannelInfo(
                    id=f"C_{channel_name.upper().replace('-', '')}",
                    name=channel_name,
                    is_private="-internal" in channel_name,
                    num_members=0,
                ))

        return channels

    def map_channel(self, channel: ChannelInfo) -> ChannelMappingResult:
        """Map a channel to its TSA context (members, roles)."""
        return ChannelMappingResult(
            channel=channel.name,
            channel_id=channel.id,
            tsa_members=[],
            eng_members=[],
            gtm_members=[],
            external_members=[],
        )

    def get_user_role(self, name: str) -> str:
        """Get the role of a user by name."""
        lower_name = name.lower().strip()
        for known_name, role in TSA_ROLES.items():
            if known_name in lower_name:
                return role
        return "unknown"


def get_channel_theme(channel_name: str) -> str:
    """Get theme for a channel."""
    if channel_name in CHANNEL_THEME_MAP:
        return CHANNEL_THEME_MAP[channel_name]["theme"]

    lower = channel_name.lower()
    for pattern, mapping in CHANNEL_THEME_MAP.items():
        if pattern.lower() in lower:
            return mapping["theme"]

    return f"Channel: #{channel_name}"


def get_channel_category(channel_name: str) -> str:
    """Get category for a channel."""
    if channel_name in CHANNEL_THEME_MAP:
        return CHANNEL_THEME_MAP[channel_name]["category"]

    lower = channel_name.lower()
    for pattern, mapping in CHANNEL_THEME_MAP.items():
        if pattern.lower() in lower:
            return mapping["category"]

    if "external" in lower or "client" in lower:
        return "customer"
    if "team" in lower or "squad" in lower:
        return "team"
    if "oncall" in lower or "ops" in lower or "incident" in lower:
        return "ops"

    return "internal"
